### ISSUE TITLE
'X' is not a know member of "None"

### ISSUE THREAD

**Issue by ChristopherZhong** [microsoft/pyright#2152]
**Describe the bug**

I have a simple one liner as follows.

```python
[x.strip() for x in os.environ.get('X').split(',') if os.environ.get('X')]
```

However, it produces the following error.

```
error: "split" is not a known member of "None" (reportOptionalMemberAccess)
```

I have tried multiple variations. The following is the `test.py` file.

```python
import os
from typing import List

def get_list_1() -> List[str]:
    value = os.environ.get('X')
    return [x.strip() for x in value.split(',')] if value else []

def get_list_2() -> List[str]:
    value = os.environ.get('X')
    return [x.strip() for x in value.split(',') if value]

def get_list_3() -> List[str]:
    return [x.strip() for x in os.environ.get('X').split(',')] if os.environ.get('X') else []

def get_list_4() -> List[str]:
    return [x.strip() for x in os.environ.get('X').split(',') if os.environ.get('X')]
```

The only one that passes is the `get_list_1` function, the rest all fails with the same error.

**To Reproduce**

Pyright version is 1.1.159.
Run `npx pyright test.py`.

**Expected behavior**

Ideally, all four variations should pass. However, I understand that `get_list_3` and `get_list_4` requires the type checker to be smart enough to know that they are the same. But at the minimum, `get_list_2` should pass?

**Comment by erictraut** [comment]
The function `os.environ.get` can return `None`. If it does, then calling `split` on the returned value will result in a runtime exception. 

`get_list_2` will also crash at runtime because `value.split(',') is executed prior to `if value`.

As you pointed out, a type checker cannot assume that a function will return the same thing each time, so `get_list_3` and `get_list_4` are not type safe.

Everything appears to be working correctly here from a type checking perspective.
